### DISCUSSION TITLE
Client Encoding Error when Connecting Database from Django

### DISCUSSION THREAD

**Discussion by nextpc24** [supabase/supabase#29065] [Questions] [answered]
# Problem
I tried to connect database through port 6543 from Django by command `python manage.py runserver` locally, but it failed. Here is the error:
```
server didn't return client encoding
psycopg2.OperationalError: server didn't return client encoding

The above exception was the direct cause of the following exception:

  File "<path-to-repo>/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "<path-to-repo>/manage.py", line 22, in <module>
    main()
django.db.utils.OperationalError: server didn't return client encoding
```
* With same packages installed and same version of python, I could connect local PostgreSQL database without issue.

## Additional Context
* Here is my installed packages from `requirements.txt`:
```
Django==5.0.6
python-dotenv==1.0.1
psycopg2==2.9.9
requests==2.32.3
djangorestframework==3.15.2
django-extensions==3.2.3
deepdiff==8.0.1
python-rapidjson==1.20
beautifulsoup4==4.12.3
django-rest-swagger==2.2.0
drf-yasg==1.21.7
sentry-sdk==2.13.0
```
* I used python 3.10.x
* I even tried to switch package between `psycopg2`, `psycopg`, and `psycopg2-binary`. But same result, except with package `psycopg` gives different error message, but I think it leads to the same error.

* Here is how I configure my database credential on `settings.py`
```
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': os.getenv('DATABASE_NAME'),
        'USER': os.getenv('DATABASE_USER'),
        'PASSWORD': os.getenv('DATABASE_PASSWORD'),
        'HOST': os.getenv('DATABASE_HOST'),
        'PORT': os.getenv('DATABASE_PORT', '5432'),
    }
}
```
* I had verified the environment variables are given and correct.
* I even had tried to switch port `5432`, but it gave me different error, which is `SSL connection has been closed unexpectedly`.
* I had tried to connect the database via `psql` and it worked, but failed on Django.

## Questions
* How do I connect database to Django without issues?
* Is there thorough blog/documentation out there for Supabase + Django integration?
* What those problems want to tell me?

**Comment by addison-grant** [comment]
Hi @nextpc24, I think the default port with Supabase is 6543. Not sure if that applies in your case.

Anyway, I'm getting that same error using SQLAlchemy + psycopg2
`OperationalError: (psycopg2.OperationalError) server didn't return client encoding`

**Reply by addison-grant** [reply]
Oh, sorry, I missed that you already tried that.

**Comment by addison-grant** [comment, marked as answer]
Hey again. I finally got past it by adding the database name `postgres` to my connection string. It looks like this:
```python
# postgresql://<host>:<password>@<db url>:<port>/<db name>
```

**Reply by addison-grant** [reply]
I also wonder if your keys under 'default' should be different and lowercase to match https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS

**Reply by nextpc24** [reply]
> Hey again. I finally got past it by adding the database name `postgres` to my connection string. It looks like this:
> 
> ```python
> # postgresql://<host>:<password>@<db url>:<port>/<db name>
> ```
nice to hear, so you meant replace the `<db_name>` with `postgres`? I would try later

**Reply by nextpc24** [reply]
> I also wonder if your keys under 'default' should be different and lowercase to match https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS

I'm not sure, I have been with Django with a same config like above for at least 3 years but never encountered this issue, but it's worth to try, I would do it and let you know the result, thanks!

**Reply by addison-grant** [reply]
> nice to hear, so you meant replace the <db_name> with postgres? I would try later

Yes. I believe the default database is called "postgres" and it has a schema called "public." I'm new to both PostgreSQL and Supabase, so I am not 100% sure I'm correct or what a schema is.

> I'm not sure, I have been with Django with a same config like above for at least 3 years but never encountered this issue, but it's worth to try, I would do it and let you know the result, thanks!

Best of luck! I'm interested to know whether it fixes it.

**Reply by nextpc24** [reply]
> Yes. I believe the default database is called "postgres" and it has a schema called "public." I'm new to both PostgreSQL and Supabase, so I am not 100% sure I'm correct or what a schema is.

Hey it works like a charm. Thanks bud!

> Best of luck! I'm interested to know whether it fixes it.

Either it's lower case or upper case, it worked, means the package can handle it either ways, so the problem is only the incorrect database name I passed to config.

**Reply by addison-grant** [reply]
Awesome to hear! 

> Either it's lower case or upper case, it worked, means the package can handle it either ways, so the problem is only the incorrect database name I passed to config.

Good to know ✍🏼
